src/shell/shell.cpp

```diff
@@ -178,141 +178,269 @@ DOS_Shell::DOS_Shell()
           call(false)
 {}
 
-// TODO: this function should be refactored to make to it easier to understand.
-// It's currently riddled with pointer and array adjustments each loop plus
-// branches and sub-loops.
-Bitu DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, bool *append)
+void DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, char **pipe, bool *append)
 {
-	char * lr=s;
-	char * lw=s;
+	char *lr = s;
+	char *lw = s;
 	char ch;
-	Bitu num=0;
 	bool quote = false;
 	char *temp = nullptr;
-	size_t temp_len = 0;
-
-	while ( (ch=*lr++) ) {
-		if(quote && ch != '"') { /* don't parse redirection within quotes. Not perfect yet. Escaped quotes will mess the count up */
+	char **redir = nullptr;
+	size_t found, temp_len;
+	std::string str, chrs;
+
+	while ((ch = *lr++)) {
+		if (quote && ch != '"') { /* don't parse redirection within
+			                     quotes. Not perfect yet. Escaped
+			                     quotes will mess the count up */
 			*lw++ = ch;
 			continue;
 		}
-
-		switch (ch) {
-		case '"':
+		if (ch == '"') {
 			quote = !quote;
-			break;
-		case '>':
-			*append=((*lr)=='>');
-			if (*append) lr++;
-			lr=ltrim(lr);
-			if (*ofn) {
-				delete[] * ofn;
-				*ofn = nullptr;
+		} else if (ch == '>' || ch == '<' || ch == '|') {
+			// Overwrite with >, and append with >>
+			if (ch == '>' && (*append = (*lr == '>')))
+				lr++;
+			redir = ch == '>' ? ofn : (ch == '<' ? ifn : pipe);
+			if (*redir) {
+				delete[] * redir;
+				*redir = nullptr;
 			}
-			*ofn = lr;
-			while (*lr && *lr!=' ' && *lr!='<' && *lr!='|') lr++;
-			//if it ends on a : => remove it.
-			if((*ofn != lr) && (lr[-1] == ':')) lr[-1] = 0;
-			temp_len = static_cast<size_t>(lr - *ofn + 1u);
-			temp = new char[temp_len];
-			safe_strncpy(temp, *ofn, temp_len);
-			*ofn = temp;
+			lr = ltrim(lr);
+			str = *redir = lr;
+			chrs = ch == '|' ? "" : (ch != '<' ? " |<" : " |>");
+			found = str.find_first_of(chrs);
+			if (found == std::string::npos)
+				temp_len = str.size();
+			else
+				temp_len = found - (str[found - 1] == ':' ? 1 : 0);
+			lr += temp_len;
+			str[temp_len] = 0;
+			temp = new char[temp_len + 1];
+			safe_strncpy(temp, str.c_str(), temp_len + 1);
+			*redir = temp;
 			continue;
+		}
+		*lw++ = ch;
+	}
+	*lw = 0;
```

Comment:
Can you add comments and help us work through the string or commands each step (or each group of steps) is parsing? 

The variable names are so short that they give no clue as to what they're doing.
The sizes, positions, temps, temp_lens, and +1 and -1's are aren't not helping me get a handle on why or what's happening.
The only hope I have to grasp onto are the few literal strings, like `" |<"` and then I know the parsing is somewhere in the middle of an `a | b` sequence.

Reply:
Sure. Added comments and made the variables names more descriptive.